### PR TITLE
Feature: Forward bridge traffic to IBC Channel

### PR DESCRIPTION
**Pull request by jackzampolin**

### CODE REVIEW

**Comment by zmanian** [comment]
there is cool.

Questions I have are

1. Are we minting and escrowing the coins on the Cosmos side of the bridge or something else? 

2. I can imagine extending the solidity contract with the send to IBC event that takes a channel directly from the eth transaction and pushes the coins over it. This allows us to preserve putting the routing tables in the client layer rather than having an unchain governance controlled routing table.

**Comment by zmanian** [comment]
Just documenting that this is timeout event handling to return funds back to the user on ethereum.

"I think we could build in some error handling where if the IBC packet times out we would send the funds back to ETH" - @jackzampolin

**Comment by jackzampolin** [comment]
1. Are we minting and escrowing the coins on the Cosmos side of the bridge or something else?
> Yes this would be the plan. Zone running gravity would mint the tokens the exact same way but instead of depositing the to desired user account it would create an IBC packet with those coins.

2. I can imagine extending the solidity contract with the send to IBC event that takes a channel directly from the eth transaction and pushes the coins over it. This allows us to preserve putting the routing tables in the client layer rather than having an unchain governance controlled routing table.
> Yup! I've actually added that in this PR too

File: module/x/gravity/keeper/ethereum_event_handler.go [hunk 1]
```diff
@@ -82,7 +86,39 @@ func (a EthereumEventProcessor) Handle(ctx sdk.Context, eve types.EthereumEvent)
...
+		}
 
+		// format of this could be {port}/{channel}
+		chanport := strings.Split(event.Channel, "/")
+
+		// TODO: here we use addr.String() we need to be able to use the acc address here encoded in the recieving chain
+		// bech32 format. We can get around this in 2 ways: one is sending it over from ETH, other is figuring out how to encode
+		// the bytes with a custom prefix which seems like it should be easy.
+		if err := a.transferKeeper.SendTransfer(ctx, chanport[0], chanport[1], coin, addr, addr.String(), clienttypes.NewHeight(0, 0), uint64(time.Second*100)); err != nil {
+			return err
```
Comment (jackzampolin) [inline module/x/gravity/keeper/ethereum_event_handler.go:118]: On this error we my want to default to just dropping the tokens in the users wallet on the gravity chain

Comment (jackzampolin) [reply module/x/gravity/keeper/ethereum_event_handler.go:118]: (i.e. in the event that channel or port doesn't exist)

**Comment by zmanian** [comment]
I feel like SendToIBCChannel should have a CosmosAddress on the gravity chain where tokens on timedout IBC packets can be sent too.

Otherwise we have a more expensive process of making a batch to send the tokens back to the eth address.

**Comment by jackzampolin** [comment]
@zmanian ETH contract emits final destination. Orchestrator parses final destination into same address w/ different prefix on both gravity and final destination chain. When event gets submitted to cosmos chain it contains both representations. IBC packets will look like this:

```json
{
  "sending_address": "cosmos1vzxkv3lxccnttr9rs0002s93sgw72h7ghukuhs",
  "reciveing_address": "osmo1vzxkv3lxccnttr9rs0002s93sgw72h7gl89vpz"
}
```

**Comment by zmanian** [comment]
makes sense

**Comment by jackzampolin** [comment]
still need to pass transfer keeper into gravity keeper

**Comment by zmanian** [comment]
In the existing SendToCosmos method, the address is encoded without the bech32 prefix and the orchestrator is configured with the correct prefixes and adds them when sending them as cosmos transactions.

Am I correct that this won't work unless we include the bech32 encoding for all the addresses?

**Comment by zmanian** [comment]
Switching to bech32 bytes seems doable.

File: module/x/gravity/keeper/ethereum_event_handler.go [hunk 2]
```diff
@@ -82,7 +86,35 @@ func (a EthereumEventProcessor) Handle(ctx sdk.Context, eve types.EthereumEvent)
...
+			}
+		}
 
+		if err := a.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, addr, coins); err != nil {
+			return err
+		}
+
+		// format of this could be {port}/{channel}
+		chanport := strings.Split(event.Channel, "/")
+		if err := a.transferKeeper.SendTransfer(ctx, chanport[0], chanport[1], coin, addr, event.ForwardAddress, clienttypes.NewHeight(0, 0), uint64(time.Second*100)); err != nil {
```
Comment (jackzampolin) [inline module/x/gravity/keeper/ethereum_event_handler.go:113]: One quick note, would be cool to add an extra parameter here to allow for gov configurable fee to be applied and funds to be put into fee pool. I think that would be pretty quick to implement too. cc @zmanian

**Comment by jackzampolin** [comment]
So RE: bech32. What I'm thinking here is that the eth transaction has the bech32 encoding for the final destination. This way the orchestrator can send that w/o having to know the prefix apriori. It can then parse that representation into bytes and also render it as the proper prefix for the gravity chain.

**Comment by EricBolten** [comment]
Looking through old PRs and this one has been sitting for almost a year. Should we plan on picking this back up or close?

**Comment by cbrit** [comment]
Closing due to staleness

**Pull request closed without merging**